crates/wasi-common/src/dir.rs

```diff
@@ -98,67 +100,83 @@ pub trait WasiDir: Send + Sync {
 }
 
 pub(crate) struct DirEntry {
-    caps: DirCaps,
-    file_caps: FileCaps,
+    caps: RwLock<DirFdStat>,
     preopen_path: Option<PathBuf>, // precondition: PathBuf is valid unicode
     dir: Box<dyn WasiDir>,
 }
 
 impl DirEntry {
     pub fn new(
-        caps: DirCaps,
+        dir_caps: DirCaps,
         file_caps: FileCaps,
         preopen_path: Option<PathBuf>,
         dir: Box<dyn WasiDir>,
     ) -> Self {
         DirEntry {
-            caps,
-            file_caps,
+            caps: RwLock::new(DirFdStat {
+                dir_caps,
+                file_caps,
+            }),
             preopen_path,
             dir,
         }
     }
     pub fn capable_of_dir(&self, caps: DirCaps) -> Result<(), Error> {
-        if self.caps.contains(caps) {
+        let fdstat = self.caps.read().unwrap();
+        Self::locked_capable_of_dir(&fdstat, caps)
+    }
+    fn locked_capable_of_dir(
+        fdstat: &impl Deref<Target = DirFdStat>,
+        caps: DirCaps,
+    ) -> Result<(), Error> {
+        let fdstat = fdstat.deref();
+        if fdstat.dir_caps.contains(caps) {
             Ok(())
         } else {
-            let missing = caps & !self.caps;
+            let missing = caps & !fdstat.dir_caps;
             let err = if missing.intersects(DirCaps::READDIR) {
                 Error::not_dir()
             } else {
                 Error::perm()
             };
-            Err(err.context(format!("desired rights {:?}, has {:?}", caps, self.caps)))
+            Err(err.context(format!(
+                "desired rights {:?}, has {:?}",
+                caps, fdstat.dir_caps
+            )))
         }
     }
-    pub fn capable_of_file(&self, caps: FileCaps) -> Result<(), Error> {
-        if self.file_caps.contains(caps) {
+    fn locked_capable_of_file(
```

Comment:
Could this perhaps be a method on `DirFdStat`?

Reply:
Yeah, sure!

Reply:
done